stan/math/prim/fun/hypergeometric_2F1.hpp

```diff
@@ -0,0 +1,207 @@
+  // // NOLINT
+  if (a1 == 1.5 && a2 == 2.0 && b == 3.0 && z < 0.0) {
+    auto abs_z = abs(z);
+    auto sqrt_1pz = sqrt(1 + abs_z);
+    return -4 * (2 * sqrt_1pz + z - 2) / (sqrt_1pz * square(z));
+  }
+
+  return {};
+}
+}  // namespace internal
+
+/**
+ * Returns the Gauss hypergeometric function applied to the
+ * input arguments:
+ * \f$_2F_1(a_1,a_2;b;z)\f$
+ *
+ * If the input parameters do not meet convergence criteria, then Euler's
+ * transformation is applied to resolve this:
+ * https://mathworld.wolfram.com/EulerTransform.html
+ *
+ * For some special-case combinations of parameters the series is calculated
+ * in closed form, see the internal::hyper_2F1_special_cases function for more
+ * details.
+ *
+ * See 'grad_2F1.hpp' for the derivatives wrt each parameter
+ *
+ * @tparam Ta1 Type of scalar first 'a' argument
+ * @tparam Ta2 Type of scalar second 'a' argument
+ * @tparam Tb Type of scalar 'b' argument
+ * @tparam Tz Type of scalar 'z' argument
+ * @param[in] a1 First of 'a' arguments to function
+ * @param[in] a2 Second of 'a' arguments to function
+ * @param[in] b 'b' argument to function
+ * @param[in] z Scalar z argument
+ * @return Gauss hypergeometric function
+ */
+template <typename Ta1, typename Ta2, typename Tb, typename Tz,
+          typename ScalarT = return_type_t<Ta1, Ta1, Tb, Tz>,
+          typename OptT = boost::optional<ScalarT>,
+          require_all_arithmetic_t<Ta1, Ta2, Tb, Tz>* = nullptr>
+inline return_type_t<Ta1, Ta1, Tb, Tz> hypergeometric_2F1(const Ta1& a1,
+                                                          const Ta2& a2,
+                                                          const Tb& b,
+                                                          const Tz& z) {
+  check_finite("hypergeometric_2F1", "a1", a1);
+  check_finite("hypergeometric_2F1", "a2", a2);
+  check_finite("hypergeometric_2F1", "b", b);
+  check_finite("hypergeometric_2F1", "z", z);
+
+  check_not_nan("hypergeometric_2F1", "a1", a1);
+  check_not_nan("hypergeometric_2F1", "a2", a2);
+  check_not_nan("hypergeometric_2F1", "b", b);
+  check_not_nan("hypergeometric_2F1", "z", z);
+
+  // Check whether value can be calculated by any special-case rules
+  // before estimating infinite sum
+  OptT special_case_a1a2 = internal::hyper_2F1_special_cases(a1, a2, b, z);
+  if (special_case_a1a2.is_initialized()) {
+    return special_case_a1a2.get();
+  }
+
+  // Check whether any special case rules apply with 'a' arguments reversed
+  // as 2F1(a1, a2, b, z) = 2F1(a2, a1, b, z)
+  OptT special_case_a2a1 = internal::hyper_2F1_special_cases(a2, a1, b, z);
+  if (special_case_a2a1.is_initialized()) {
+    return special_case_a2a1.get();
+  }
+
+  Eigen::VectorXd a_args(2);
+  Eigen::VectorXd b_args(1);
+
+  try {
+    check_2F1_converges("hypergeometric_2F1", a1, a2, b, z);
+
+    a_args << a1, a2;
+    b_args << b;
+    return hypergeometric_pFq(a_args, b_args, z);
+  } catch (...) {
```

Comment:
idt we normally use ..., are there several types of errors that can be thrown?  I think our usual is to catch `const std::exception&` and if we don't want to do anything with it we leave it unnamed